File: assignment 6/c/programs/4.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 100

char stack[MAX_SIZE];
int top = -1;

void push(char element)
{
   if (top == MAX_SIZE - 1)
   {
      printf("Stack overflow\n");
      return;
   }
   stack[++top] = element;
}

char pop()
{
   if (top == -1)
   {
      printf("Stack underflow\n");
      return -1;
   }
   return stack[top--];
}

int precedence(char operator)
{
   if (operator== '+' || operator== '-')
   {
      return 1;
   }
   else if (operator== '*' || operator== '/')
   {
      return 2;
   }
   else if (operator== '^')
   {
      return 3;
   }
   else
   {
      return -1;
   }
}

void infixToPostfix(char *infixExpression, char *postfixExpression)
{
   int length = strlen(infixExpression);
   int j = 0;
   for (int i = 0; i < length; i++)
   {
      char currentCharacter = infixExpression[i];
      if (currentCharacter >= 'a' && currentCharacter <= 'z')
      {
         postfixExpression[j++] = currentCharacter;
      }
      else if (currentCharacter == '(')
      {
         push(currentCharacter);
      }
      else if (currentCharacter == ')')
      {
         while (stack[top] != '(')
         {
            postfixExpression[j++] = pop();
         }
         pop();
      }
      else
      {
         while (top != -1 && precedence(currentCharacter) <= precedence(stack[top]))
         {
            postfixExpression[j++] = pop();
         }
         push(currentCharacter);
      }
   }
   while (top != -1)
   {
      postfixExpression[j++] = pop();
   }
   postfixExpression[j] = '\0';
}

int main()
{
   char infixExpression[MAX_SIZE];
   printf("Enter an infix expression: ");
   scanf("%s", infixExpression);
   char postfixExpression[MAX_SIZE];
   infixToPostfix(infixExpression, postfixExpression);
   printf("The postfix expression is: %s\n", postfixExpression);
   return 0;
}

File: assignment 6/python/programs/4.py
def infix_to_postfix(expression):
    # Define the set of operators and their priorities
    operators = set(['+', '-', '*', '/', '(', ')', '^'])
    priority = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}

    # Initialize an empty stack and output string
    stack = []
    output = ''

    # Iterate through each character in the expression
    for character in expression:
        if character not in operators:
            # If the character is an operand, add it to the output string
            output += character
        elif character == '(':
            # If the character is an opening parenthesis, push it onto the stack
            stack.append('(')
        elif character == ')':
            # If the character is a closing parenthesis, pop operators from the stack and add them to the output string until an opening parenthesis is found
            while stack and stack[-1] != '(':
                output += stack.pop()
            stack.pop()
        else:
            # If the character is an operator, pop operators from the stack and add them to the output string until an operator with lower priority is found
            while stack and stack[-1] != '(' and priority[character] <= priority[stack[-1]]:
                output += stack.pop()
            stack.append(character)

    # Pop any remaining operators from the stack and add them to the output string
    while stack:
        output += stack.pop()

    return output

# Example usage
expression = input('Enter an infix expression: ')
print('Infix expression:', expression)
print('Postfix expression:', infix_to_postfix(expression))
